src/atomate2/qchem/sets/base.py

```diff
@@ -0,0 +1,157 @@
+"""Module defining base QChem input set and generator."""
+
+import os
+from pickletools import optimize
+import warnings
+from copy import deepcopy
+from dataclasses import dataclass, field
+from itertools import groupby
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import numpy as np
+from monty.io import zopen
+from monty.serialization import loadfn
+from pkg_resources import resource_filename
+from pymatgen.core.structure import Molecule
+from pymatgen.io.qchem.inputs import QCInput
+#from pymatgen.io.qchem.utils import lower_and_check_unique
+from pymatgen.io.core import InputGenerator
+from pymatgen.io.qchem.sets import QChemDictSet
+from atomate2 import SETTINGS
+
+__author__ = "Alex Ganose, Ryan Kingsbury, Rishabh D Guha"
+__copyright__ = "Copyright 2018-2022, The Materials Project"
+__version__ = "0.1"
+
+_BASE_QCHEM_SET = loadfn(resource_filename("atomate2.qchem.sets", "BaseQchemSet.yaml"))
+
+__all__ = ["QChemInputGenerator"]
+
+@dataclass
+class QChemInputGenerator(InputGenerator):
+    """
+    A class to generate QChem input sets for different calculations
+    """
+
+    def get_input_set( 
+        self,
+        molecule: Molecule = None,
+        prev_dir: Union[str, Path] = None,
+        overwrite_inputs: dict = None
+    ) -> QChemDictSet:
+        """
+        Get a QChem Input Set as a dictionary for a molecule
+
+        Parameters
+        ----------
+        molecule
+            A Pymatgen molecule
+        prev_dir
+            A previous directory to generate the input set from
+        overwrite_inputs
+
+        Returns
+        -------
+        QchemInputSet
+            A QChem input set
+        """
+        molecule, prev_basis, prev_scf, new_geom_opt, overwrite_inputs, nbo_params = self._get_previous(
+            molecule, prev_dir
+        )
+
+        basis_set_updates = self.get_basis_set_updates(
+            molecule,
+            prev_basis=prev_basis,
+            prev_scf=prev_scf,
+            new_geom_opt=new_geom_opt,
+            overwrite_inputs=overwrite_inputs,
+            nbo_params=nbo_params,
+        )
+
+        scf_algorithm_updates = self.get_scf_algorithm_updates(
+            molecule,
+            prev_basis=prev_basis,
+            prev_scf=prev_scf,
+            new_geom_opt=new_geom_opt,
+            overwrite_inputs=overwrite_inputs,
+            nbo_params=nbo_params,
+        )
+
+        basis_set = self._basis_set(basis_set_updates)
+        scf_algorithm = self._get_scf_algorithm(scf_algorithm_updates)
+
+
+        return QChemDictSet(
```

Comment:
Can we not do this. It will make it very challenging to diagnose why certain things are getting written in the input sets. You will have to first check `QChemInputGenerator`, then check `QChemDictSet` and then finally check `QCInput`.

We should only have the generator `QChemInputGenerator` and the input set `QCInput`. The long term goal is to use only the modern input sets and remove the old style input sets. This won't be possible if the new input sets rely on the old ones.

So please can you refactor `QChemDictSet` into `QChemInputGenerator`. It does not look like it will take that long as the QChem inputs are a lot simpler than the VASP input sets.

Reply:
@utf I completely understand the sentiment and agree that refactoring `QChemDictSet` into a modern `InputSet` is the long term goal. When I looked into doing that before though, it did not seem straightforward. So I had suggested to Rishabh that as long as the legacy `QChemDictSet` behaves the same way as a modern `InputSet` (e.g., has a `write_input` method with the correct call signature), a sensible incremental solution would be to use an `InputGenerator` to create a legacy `InputSet`.

In general, I feel that the QChem `InputSet` do a lot less overrriding of input kwargs than the VASP ones so I don't think there's as high a risk of this becoming a confusing mess. @rdguha1995 it would be good to see if you can skip over the `DictSet` layer and directly create `QCInput` here, which would reduce the risk of a kwarg getting overridden somewhere. But I'm not sure how well that will work because I think the existing `QCInput` is somewhere in between the functionality of a modern `InputSet` and `InputFile`.

Aside from the risk of confusing overrides, I don't believe taking this approach adds any extra dependencies or complexifies the code (but please correct me if I'm wrong). If that's the case, I'd propose that we use this paradigm to get to beta quality atomate2 support for Q-Chem and tackle the `InputSet` refactor a bit down the line, just to minimize the number of moving parts.